openff/toolkit/utils/rdkit_wrapper.py

```
@@ -2298,137 +2299,225 @@ def _detect_undefined_stereo(cls, rdmol, err_msg_prefix="", raise_warning=False)
-
-            Chem.AssignStereochemistry(rdmol, cleanIt=True, force=True)
-
-            # Verify that the current directions give us the desired stereochemistries.
-            assert bond.stereochemistry in {"E", "Z"}
-            if bond.stereochemistry == "E":
-                desired_rdk_stereo_code = Chem.rdchem.BondStereo.STEREOE
-            else:
-                desired_rdk_stereo_code = Chem.rdchem.BondStereo.STEREOZ
+        # Do not assume that every solution found by the solver is valid.
+        # Iterate through the solutions and ensure that the
+        # desired E/Z marks have really been achieved.
+        has_solution = False
 
-            # If that doesn't work, flip the direction of one bond preferring
-            # those that are not already determining the stereo of another bond.
-            if stereo_rdbond.GetStereo() != desired_rdk_stereo_code:
-                cls._flip_rdbond_direction(flipped_rdbond, paired_bonds)
-                Chem.AssignStereochemistry(rdmol, cleanIt=True, force=True)
+        for solution in csp_problem.getSolutionIter():
 
-                # The stereo should be set correctly here.
-                assert stereo_rdbond.GetStereo() == desired_rdk_stereo_code
+            for rd_bond_index, direction in solution.items():
 
-            # Update paired bonds map.
-            neighbor_bond_indices = [
-                (rdb.GetBeginAtomIdx(), rdb.GetEndAtomIdx()) for rdb in neighbor_rdbonds
-            ]
-            for i, bond_indices in enumerate(neighbor_bond_indices):
-                try:
-                    paired_bonds[bond_indices].append(neighbor_rdbonds[1 - i])
-                except KeyError:
-                    paired_bonds[bond_indices] = [neighbor_rdbonds[1 - i]]
+                rd_bond = rd_molecule.GetBondWithIdx(rd_bond_index)
+                if direction == 0:
+                    rd_bond.SetBondDir(Chem.BondDir.ENDDOWNRIGHT)
+                elif direction == 1:
```

In general you shouldn't have a dangling `elif` statement as it can let unexpected values silently slip through. This should either be an `else` or you should add an `else: raise NotImplementedError()` to handle the unexpected edge case.

---

Good call. Fixed in [86df78d](https://github.com/openforcefield/openff-toolkit/pull/1231/commits/86df78d062269a6ab8bf4658ba5072e2a51fc85e)